### ISSUE TITLE
Add code to avert 5000 list item limit

### ISSUE THREAD

**Issue by russgove** [OfficeDev/PnP-JS-Core#115]
Thank you for reporting an issue, suggesting an enhancement, or asking a question. We appreciate your feedback - to help the team understand your
needs please complete the below template to ensure we have the details to help. Thanks!

#### Category
[x ] Enhancement

[ ] Bug

[ ] Question

We should implement something to overcome/avert the 5000 item limit imposed by sharepoint. Maybe get batches of 5000 in sub-reqauest, but don't resolve the final promise until we have them all.
see http://sympmarc.com/2016/05/24/dear-microsoft-remove-the-5000-item-limit-for-sharepoint-list-views/

**Comment by mike-morrison** [comment]
I believe the most you can request from the REST API is 2000 at a time, but that 
was a while ago.  Also it was search, and trying to take the full 2000 took too long.  It took too long for the server to return the first byte.  It took too long for the 2000 search results to come back to the browser, and due to lack of threading, it kept the UI thread locked while popping the 2k results on the screen which also took a while.  

There is no 5000 item limit in SharePoint.  The throttle is based off of operations, not items.  For example, if I am showing a list of 10,000 items.  I can page through all 10000 of them, 50 at a time quick and easy.  If I try to add a "group by" to this list, this now causes a full list read to place everything into groups.  Now you will hit that 5000 operation limit without adding a single item.  We support up into the millions for list items.  As for why this limitation didn't exist in 2007 vs 2010+.  I happen to know part of the answer to this.

#### Boring Background (feel free to skip)

Back when 2007 was in mainstream support, we got a LOT of performance cases.  The number of cases we used to get where pages wouldn't load, or when the entire farm slowed down to a crawl was too high.  We almost always traced this back to a large list or a faulty full trust code solution.  

This happened so frequently, that we had SQL scripts that would help us identify these large lists.  What the script really checked was that there was no more than 2000 items per view.  Meaning no more than 2000 total items in a folder or view.  

We were constantly having to explain to customers that SQL was spending too many resources on a few lists and not leaving resources for the rest of the farm.  The only way out was to restructure the list.  Usually archiving by Year/Month, etc.  

One of the many things the Product Group looks at when adding "features" to the product, is how we can reduce the number of support incidents a particular part of the product is generating.  This list throttle "feature" was added to mitigate the customer pain point of inefficient SQL queries caused by lists not being properly organized or indexed.

Keep in mind, a list in SharePoint is not a thin layer on top of SQL.  It is a complex application, built on top of SQL.  It is focused on doing small to medium size end user list scenarios. Think employee in and out boards, not a place to store event logs from a server.  Although you can do the latter, is a SharePoint list really the right place for this?  

In that article it is mentioned that Access is much faster.  In almost every scenario, that would be accurate, they are 2 different products doing 2 different things.  It's an unfair comparison.  

Access does not have to security trim the results, or perform many of the per item operations that SharePoint has to perform like hierarchical content types, workflows, etc.  This makes it an inefficient data platform for a lot of scenarios.  SharePoint is "one size fits most" and it does it's job well.  Access Services fills some of this gap, being a thinner layer on top of SQL tables.  

@sympmarc is right in being frustrated and saying this needs to be fixed.  Things like counts, sums, reporting, etc. should just work.  A lot of times the feature feels more like a roadblock.
  
#### TLDR;

Honestly, the real bottleneck is JavaScript's single thread.  We could add a LOT of value here.  All of my testing (and common sense) has shown that the REST API is many times faster in .NET where I can Parallel.ForEach()  on background threads which is essentially what web workers give us.  

Browser support is pretty good, IE 10+, in other browsers for longer.  At least one of the libraries I have seen in the past will fallback to promises in older browsers.  If we could make this dead simple, or better transparent.  That would be huge.  Right now, it is a big negative for me for this library.  With REST directly, I can use web workers fairly easily.  With this library, I am not sure how I could decouple it and make it web worker friendly.

**Comment by russgove** [comment]
Thanks for the detailed reply Mike. I tried to do some testing to see how/if I could get all items from a list with our library. I created a list with 5000 items. When I issue a get on the list , rest returns me only 100 items. According to the docs , there is supposed to be a 'nextLink' property in the response to get the  next batch of items, I looks like this property is not being returned by the Queryable defaultParser, so I cannot use it to get the next batch of items.

The screenshot attached shows that sharepoint is returning the nextLink to get the next batch of items, This link is not being passed back to the caller though, We're just returning the array of 100 items. 

I'll open an issue for us to return the nextLink, maybe as an attribute on the array, but I'm not sure of the implications.
![capture](https://cloud.githubusercontent.com/assets/2412956/15636812/80989d86-25d8-11e6-845a-017338acbb5e.PNG)

**Comment by sympmarc** [comment]
@russgove: The default is 100 (which is almost never enough). You can specify $top=nnnn, where nnnn is up to 5000; above that gives you an error. Using $skip=nnnn means 'start with item nnnn + 1'.

**Comment by sympmarc** [comment]
@russgove: p.s. To cover all situations, I think you should return all name/value pairs that come back from the request in an array, along with the data. One of the problems with wrapping something like this is it can obscure details that matter in situations you can't predict. Trying to know what all those are can be a fool's errand.

**Comment by sympmarc** [comment]
@mike-Morrison: There is indeed a limit of 5000 items in the REST API. Try it. 2000 was the "recommended" maximum list size in 2007; the point at which it was deemed a "large list". But we could query as many items as we wanted. The 5000 item limit became a hard one with the REST interface in 2010. It's not based on operations; it's a hard number. It's simply not enough for real applications much of the time. 

As for your comments about how long it takes to get the data, that's a serious "it depends". The "width" of the request matters as much as the length. We have to be smart with our requests, just as you have to be on the server. If you don't use a $select to only retrieve what you need, then you're not being efficient. I see server side devs do this sort of thing as often as client side devs. It's a matter of education and skill, and limits shouldn't be forced on us to cover all potential dumb requests. Issue warnings; optimize calls; whatever. But don't make the platform deficient or people will go elsewhere.

**Comment by russgove** [comment]
Leave a comment

**Comment by patrick-rodgers** [comment]
@russgove, I have added support for paging through items using the getPaged method added to the items collection in PR #119. This now returns a typed object that has the data in a "results" property and a method getNext that will get the next set based on the initial query. You can see a sample of using it below:

```
    pnp.sp.web.lists.getByTitle("Config3").items.orderBy("Title").top(2).getPaged().then(d => {
        show(d);
        d.getNext().then(d => show(d));
    });    
```

In terms of providing a "getAll" method, I am against that because it has real potential to cause serious issues with performance. One of the largest issues we see from customers when they complain that things are slow in O365 is poor optimization of queries from the client. Often a list will start out with 3 things during testing then suddenly have 100s, think news articles. So when a developer first used an equivalent of getAll it was fine - a few months later the site is unresponsive because a web part on the home page for all users company wide is pulling 150 articles to show 2. 

Using the getPaged method we now support paging for list items, which I agree was missing as you noted so good catch on that. We need to remain focused on getting the least number of items on the client side.

@sympmarc, on the point of wrapping - some folks have been very vocal the other way - and in most cases I tend to agree. If I do an items.get() I likely want then to just loop that array. Something we are moving closer supporting through increased typing as seen in WIP #113. Perhaps we need to include a ODataRaw parser that just dumps you the raw output with no parsing. Always trying to find that balance between usability and hiding too much.

**Comment by russgove** [comment]
That's Great! Thanks Patrick.
Regarding the getAll... This is needed not for user pages but for 'database maintenance' type routines where one needs to 'visit' every record in a table to perform some operation.  If you don't want to add the method, I guess we can just give a sample of how to do it via getPaged.

Thanks for the quick response.

Russell

**Comment by sympmarc** [comment]
@russgove Sometimes getAll *is* required. Now that we're going to expect people to do much of their dev client side, we'll need to cover all the stuff they are doing today. The simplest of examples is summing some value by the value of another column: sales by department, etc.

**Comment by patrick-rodgers** [comment]
@sympmarc, let's discuss it on the call. It is easy enough to write a getAll method, but what worries me is knowing how people write code in the real world it will become the "default" method. So we would essentially be introducing performance problems. Would some people use it responsibly? Of course. I'm not worried about them, I'd worried about the rest. Back to my scenario of a web part on the corp homepage that pulls all articles from a list with 1000s of items with 50 columns (I've seen it) unfiltered then sorts client side and shows the top 3.

You could also say that so what, they made the mistake and we would just be providing the functionality. I could also see coming up with an iterator type implementation where you would provide a function and get a set of items from the list so that the library wouldn't ever hold all the items in memory at a time but would allow for things like summing some value or other aggregate type operations. I'd probably lean more that way than a true "getAll".

**Comment by russgove** [comment]
Patrick, The callback method is perfect for the database maintenance type of routine I mentioned above. 
I agree that novice developers would see a getAll method and just use that which could lead to performance problems.  We can either leave it out , and provide a sample on how to do it(which may scare off novice developers), or include it (with documentation that it can perform poorly when etc....).

**Comment by mike-morrison** [comment]
@sympmarc The documentation explains it better than I do, the following excerpt is sourced [from here](https://msdn.microsoft.com/en-us/library/ff798465.aspx).   I have seen a few scenarios where you get throttled on lists with less than 5000 items.

>
#### How Does Indexing Affect Throttling?
> 
The list view threshold does not apply simply to the number of results returned by your query. Instead, it restricts the numbers of database rows that can be accessed in order to complete execution of the query at the row level in the content database. For example, suppose you are working with a list that contains 10,000 items. If you were to build a query that returns the first 100 items sorted by the ID field, the query would execute without issue because the ID column is always indexed. However, if you were to build a query that returns the first 100 items sorted by a non-indexed Title field, the query would have to scan all 10,000 rows in the content database in order to determine the sort order by title before returning the first 100 items. Because of this, the query would be throttled, and rightly so—this is a resource-intensive operation.

I don't look at width for optimizing server performance.  I care about width because it increases length.  If I had a very wide list (> 80 columns) or a list with a few lookup columns, I might use $select to exclude unnecessary lookups a.k.a. SQL joins to speed that up on the server.  However, the main reason I use $select is to reduce the payload (length) being returned.  The slight server performance increase is just gravy.  That's all just my opinion, though.  

##### @Everyone
I would prefer we do not add a getAll.  I agree with @patrick-rodgers, it would likely become the default when IMO, it should only be used for edge cases.  At that point we are promoting an anti-pattern which is requesting more data than you need, while at the same time adding a performance hit and memory tax to most interactions.  

I strongly believe that we should keep the lightweight and flexible goodness that is inherent in the REST API.  I could see more benefit in caching the next and previous items via Boolean switches, or making web workers automatic if available.

**Comment by patrick-rodgers** [comment]
Going to close this. We've now got support for list item paging and I added iterating list items to the roadmap page in the wiki (a new page!). Thanks for the great discussion and if I missed anything feel free to re-open this. Thanks!
